conda/core/solve.py

```diff
@@ -1324,28 +1347,93 @@ def _configure_solver_for_remove(self, state):
+        if is_update and update_modifier == UpdateModifier.UPDATE_ALL:
+            tasks["SOLVER_UPDATE"].update(
+                {
+                    pkg.name: MatchSpec(pkg.name).conda_build_form()
+                    for pkg in state["installed_pkgs"]
+                    if not pkg.is_unmanageable
+                }
+            )
+
+        # C. Implicitly requested via `install --update-deps`
         elif update_modifier == UpdateModifier.UPDATE_DEPS:
-            final_specs = specs_to_add
-            for spec in specs_to_add:
-                prec = next(p for p in state["installed_pkgs"] if p.name == spec.name)
-                for dep in prec.depends:
-                    ms = MatchSpec(dep)
-                    if ms.name != "python":
-                        final_specs.append(MatchSpec(ms.name))
-            specs_to_add += list(set(final_specs))
-        return [s.conda_build_form() for s in specs_to_add]
+            # This code below is adapted from the legacy solver logic
+            # found in Solver._post_sat_handling()
+
+            # We need a pre-solve to find the full chain of dependencies
+            # for the requested specs (1) This first pre-solve will give us
+            # the packages that would be installed normally. We take
+            # that list and process their dependencies too so we can add
+            # those explicitly to the list of packages to be updated.
+            # If additionally DEPS_ONLY is set, we need to remove the
+            # originally requested specs from the final list of explicit
+            # packages.
+
+            originally_requested_specs = self.specs_to_add
+            with context.override("quiet", True):
+                # (1) Presolve to find the dependencies for `specs_to_add` (CLI)
+                solved_pkgs = self.solve_final_state(
+                    update_modifier=UpdateModifier.UPDATE_SPECS,
+                    deps_modifier=deps_modifier,
+                    ignore_pinned=ignore_pinned,
+                    force_remove=force_remove,
+                    force_reinstall=force_reinstall)
+            specs_to_add_plus_deps = self.specs_to_add
+
+
+            # (2) Get the dependency tree for each dependency
+            graph = PrefixGraph(solved_pkgs)
+            update_names = set()
+            for spec in specs_to_add_plus_deps:
+                node = graph.get_node_by_name(spec.name)
+                update_names.update(ancest_rec.name for ancest_rec in graph.all_ancestors(node))
+            specs_map = {name: MatchSpec(name) for name in update_names}
+
+            # Remove pinned_specs and any python spec (due to major-minor pinning business rule).
+            for spec in self._pinned_specs():
+                specs_map.pop(spec.name, None)
+            if "python" in specs_map:
+                python = next((p for p in state["installed_pkgs"] if p.name == "python"), None)
```

Comment:
Let's add a TODO item here that we may want to abstract this into a utility API for other cases in which we want to restrict this spec map.